lib/form/paragraph3.flow

```diff
@@ -630,160 +674,155 @@ keepFormTogether(form : Form) -> bool {
+ParaLineMetrics(width : double, height : double, asc : double);
 OptimizedLineElement(f : GhostForm, metrics : Behaviour<FormMetrics>, interactivityIdM : Maybe<int>);
 
-getOptimizedLine(words : [ParaWord], alignment : ParaLineAlignment) -> [OptimizedLineElement]{
-	if (!isBiDiEnabled() && alignment != Justify()) {
+getOptimizedLine(words : [ParaWord], alignment : GeneralAlignment) -> [OptimizedLineElement]{
+	if (!isBiDiEnabled() && equalDoubles(alignment.justify, 0.0)) {
 		optimizeLine(words);
 	} else {
 		map(words, \w -> OptimizedLineElement(w.form, w.metrics, w.interactivityIdM));
 	};
 }
 
+measureLine(words : [OptimizedLineElement], lineIndent: double) -> ParaLineMetrics {
+	metrics = map(words, \p -> getValue(p.metrics));
+	lineAsc = fold(metrics, 0.0, \ac, m -> max(ac, m.baseline));
+	ParaLineMetrics(
+		lineIndent + fold(metrics, 0.0, \ac, m -> ac + m.width), // this is real width of the line at the moment.
+		fold(metrics, 0.0, \ac, m -> max(ac, lineAsc + m.height - m.baseline)),
+		lineAsc
+	)
+}
+
 // Aligns all elements to the baseline, as well as calculates the height and the baseline
 // Send out the result, width, height
 RenderLine(
 	y : double,
-	words : [ParaWord],
+	optimizedLine : [OptimizedLineElement],
 	width : double,
-	tightWidth : bool,
-	alignment : ParaLineAlignment,
+	lineMetrics: ParaLineMetrics,
+	alignment : GeneralAlignment,
 	lineIndent: double,
 	lineNumber : int,
 	interactiveStyles : Tree<int, [ParaElementInteractiveStyle]>,
 	rtl : bool,
 	isMarked
-) -> ParaLineResult {
-	optimizedLine = getOptimizedLine(words, alignment);
-	metrics = map(optimizedLine, \p -> getValue(p.metrics));
-	lineWidth = lineIndent + fold(metrics, 0.0, \ac, m -> ac + m.width); // this is real width of the line at the moment.
-	lineAsc = fold(metrics, 0.0, \ac, m -> max(ac, m.baseline));
-	lineHeight = fold(metrics, 0.0, \ac, m -> max(ac, lineAsc + m.height - m.baseline));
-
+) -> Form {
 	// Alignment offset
-	startOffset = if (rtl) max(width - lineWidth, 0.) else 0.;
-	endOffset = if (rtl) 0. else max(0., width - lineWidth);
-	alignmentOffset = ref switch (alignment : ParaLineAlignment) {
-		LeftAlign(): if (rtl) endOffset else startOffset;
-		RightAlign(): if (rtl) startOffset else endOffset;
-		CenterAlign(): max(0.0, (width - lineWidth) / 2.0);
-		StartAlign(): startOffset;
-		EndAlign(): endOffset;
-		Justify(): 0.0;
-	} + (if (rtl) 0. else lineIndent);
+	dWidth = max(0.0, width - lineMetrics.width);
+	startOffset = dWidth*alignment.start + lineIndent;
 	// Inter-word spacing for justification
-	interWord = if (alignment == Justify() && length(optimizedLine) > 1) {
-		nwords = length(optimizedLine);
-		remaining = (width - lineWidth);
-		min(10.0, remaining / i2d(nwords - 1));
-	} else 0.0;
-
+	nwords = length(optimizedLine);
+	interWord = if (nwords <= 1) 0.0 else
+		min(10.0, dWidth * alignment.justify / i2d(nwords - 1));
+	x = ref if (rtl) dWidth-startOffset-interWord*i2d(nwords-1) else startOffset;
```

Comment:
please, use spaces around arithmetic signs, it's hard to read such expressions

Reply:
and this expression is hard to understand in anyway. looks like a mess

Reply:
I will.